### PR TITLE
Fix opcode `EXTCODEHASH` to use non-existing proof for account existence

### PR DESCRIPTION
**Pull request by silathdiir**
According to previous `BALANCE` PR https://github.com/privacy-scaling-explorations/zkevm-specs/pull/248, fix this opcode to use non-existing proof for account existence check.

### CODE REVIEW

**Comment by silathdiir** [comment]
Hi @ed255 , could you help review this PR when you have time? It refactors opcode `EXTCODEHASH` to use `AccountFieldTag.NonExisting` as previous `BALANCE` PR #248 . Thanks.

File: src/zkevm_specs/evm/execution/extcodehash.py [hunk 1]
```diff
@@ -1,37 +1,38 @@
...
-        * instruction.is_zero(balance)
-        * instruction.is_equal(code_hash, instruction.rlc_encode(EMPTY_CODE_HASH, 32))
-    )
+    if exists == 0:  # 1 - exists == 1
+        instruction.account_read(address, AccountFieldTag.NonExisting)
 
+    code_hash = (
+        instruction.account_read(address, AccountFieldTag.CodeHash)
+        if exists == 1
+        else instruction.rlc_encode(EMPTY_CODE_HASH, 32)
```
Comment (ed255) [inline src/zkevm_specs/evm/execution/extcodehash.py:25]: maybe this `else` is unneeded: when `exists == 0`, the `code_hash` value will not be used.

File: tests/evm/test_extcodehash.py [hunk 2]
```diff
@@ -1,74 +1,69 @@
...
 
     code_hash = int.from_bytes(keccak256(code), "big")
-    result = 0 if (nonce == 0 and balance == 0 and code_hash == EMPTY_CODE_HASH) else code_hash
+    result = code_hash if exists == 1 else 0
 
     tx_id = 1
     call_id = 1
 
-    rw_counter_end_of_reversion = 0 if is_persistent else 9
+    # 9 + 1 reversible operation (the account access list write)
```
Comment (ed255) [inline tests/evm/test_extcodehash.py:62]: Isn't this `7 + 1`?

**Review by ed255 — APPROVED** [review]
LGTM! I've added two minor notes.

File: src/zkevm_specs/evm/execution/extcodehash.py (hunk 1 above)
Comment (silathdiir) [reply src/zkevm_specs/evm/execution/extcodehash.py]: Fixed to only set `code_hash` when `exists == 1` as:
```
if exists == 1:
    code_hash = instruction.account_read(address, AccountFieldTag.CodeHash)
else:  # exists == 0
    instruction.account_read(address, AccountFieldTag.NonExisting)
```
And also updated code in `balance.py`.

File: tests/evm/test_extcodehash.py (hunk 2 above)
Comment (silathdiir) [reply tests/evm/test_extcodehash.py]: Fixed to `7 + 1`.

File: src/zkevm_specs/evm/execution/extcodehash.py (hunk 1 above)
Comment (silathdiir) [reply src/zkevm_specs/evm/execution/extcodehash.py]: I am sorry that I missed `make install` to test this code 😂. Cell `code_hash` must be assigned. I reverted the code as:
```
if exists == 0:
    instruction.account_read(address, AccountFieldTag.NonExisting)

code_hash = (
    instruction.account_read(address, AccountFieldTag.CodeHash) if exists == 1 else RLC(0)
)
```
Similar as `BALANCE`.

Comment (ed255) [reply src/zkevm_specs/evm/execution/extcodehash.py]: ah no problem! This looks good :)

**Review by roynalnaruto — APPROVED** [review]
Looks good to me

**Pull request merged**